### DISCUSSION TITLE
Trigger "Sync last n workouts" from Home Assistant

### DISCUSSION THREAD

**Discussion by geoffbon** [philosowaffle/peloton-to-garmin#691] [Ideas]
I'm using Home Assistant, in conjunction with the [Garmin HA app](https://github.com/house-of-abbey/GarminHomeAssistant), and so have the ability to detect when I've completed a workout, at least within whatever refresh interval I set for the Garmin's app background service.

Ideally I'd like to have an automation in HA detect when an activity has finished, and call an endpoint to sync the last 1 (or n) workouts.  
Or, something I can tap on my HA tablet dashboard to trigger the sync.  

I see one API endpoint to sync a specific workout ID, and another endpoint to get the last n ids, but this is going to add some serious complexity to my simple automation.  Is it feasible to expose an API call that just takes an integer as a URI parameter to trigger a sync of that many most recent workouts?  Or even just a parameterless call that syncs the most recent workout?

With that, I can just turn off the background sync in the app, and get more timely sync when I need it.

**Comment by philosowaffle** [comment]
Makes sense, created #697 to get it on the roadmap.

**Comment by adolphe222** [comment]
Is there any documentation on how to use the API from something like Home Assistant.  I'm pretty new to doing something like that.  Sounds like it's now possible for HA integration, but I've been reading the wiki and can't figure out where to start. 

My goal is I would like to have a wireless button on my peloton that I can press when my workout is complete to trigger the sync. My though it is to setup the button with HA and then have HA trigger the sync.

**Reply by philosowaffle** [reply]
No proper documentation for this, but if you have the API already running then you can visit `http://<serverIP>:<port>/swagger` to see the entire API spec and play around with it.

The endpoint you'll most likely want to use is `POST api/sync/recent` which accepts a body of:
```json
{
  "numberToSync": 0,
  "forceStackWorkouts": true
}
```

The P2G api is currently not protected by any kind of authentication.

Not sure your familiarity with these kinds of things, but if you're hosting P2G using the sample docker compose file, you'll want to uncomment [these lines ](https://github.com/philosowaffle/peloton-to-garmin/blob/master/docker/webui/docker-compose-ui.yaml#L13) to ensure you can interact directly with the API.

**Reply by adolphe222** [reply]
Thank you for the speedy response.  Definitely a lot of learning for me.  I was able to get to the swagger site and see the API.  I'm still figuring out if I have things right on the Home Assistant end since they ask for everything to be in YAML.

**Reply by adolphe222** [reply]
Thanks again for your help.  I got Home Assistant Rest command working correctly.  My main issue seemed to be that when I used
```
{
  "numberToSync": 0,
  "forceStackWorkouts": true
}

```
I was getting a 400 response.  I set "numberToSync" to 1 and then it works.  If I want to sync all recent workouts from today (or unsynced workouts) will using "1" work? Or any recommendation how to handle?  My plan is to able to trigger the sync through Home Assistant when done the workout. Sometimes I do a single workout, and sometimes I do multiple like a warmup, then a class, and then a cool down.

**Reply by philosowaffle** [reply]
No, `numberToSync` is exactly that, the number of most recent workouts to sync.  If you set it to `0` it syncs nothing (well, throws you an error, same thing).  If you set it to `1` it will sync the single most recent workout.  If you set it to `10` it will sync the ten most recent workouts. 

I would choose a number generally guaranteed to be higher than the number of workouts you do.  If you easily get 5 workouts in a session, then choose `5` for example.  This will ensure all of your recently completed workouts get synced, and Garmin already rejects duplicate syncs so you don't need to worry about re-syncing something that got synced yesterday.

But honestly, for the best experience, you could probably make this number a [custom helper](https://www.home-assistant.io/integrations/template/#number) in Home Assistant, then you could adjust it to the correct number from your phone before syncing.  This will also yield the best result for the next paragraph.....

👀  If you don't use P2G workout stacking at all then you can ignore all of the below, just be sure to set `forceStackWorkouts=false`.

You probably want to think about the value for `forceStackWorkouts` as well.   [Here there be dragons.](https://github.com/philosowaffle/peloton-to-garmin/issues/753#issuecomment-2745276604) especially when `numberToSync` ends up grabbing just 1 or 2 workouts from the previous day.  This CAN result in duplicates getting uploaded to Garmin.

Here's the [general documentation](https://philosowaffle.github.io/peloton-to-garmin/latest/configuration/format/#stacked-workouts) for this feature.
